weblate/trans/models/change.py

```diff
@@ -114,14 +115,26 @@ def prefetch(self):
 
         Call prefetch or prefetch_list later on paginated results to complete.
         """
+        translation = apps.get_model("trans", "Translation")
+        component = apps.get_model("trans", "Component")
+        unit = apps.get_model("trans", "Unit")
         return self.prefetch_related(
             "user",
             "author",
-            "translation",
-            "component",
+            Prefetch(
+                "translation",
+                queryset=translation.objects.exclude(component__to_delete=True),
+            ),
+            Prefetch(
+                "component",
+                queryset=component.objects.exclude(category__to_delete=True),
+            ),
             "project",
             "component__source_language",
-            "unit",
+            Prefetch(
+                "unit",
+                queryset=unit.objects.exclude(translation__component__to_delete=True),
+            ),
```

Comment:
This is unnecessarily complex, better to filter changes for objects not scheduled for removal.